### PR TITLE
Implement zilliqa observer

### LINKED ISSUE #208: [Observer] Add Zilliqa support for BlockAPI
Docs: https://apidocs.zilliqa.com/#getrecenttransactions

### PR DESCRIPTION
**Pull request by hewigovens**
Fixes #208, changes: 
- Add another json rpc client
- Get tx block height
-  Send batch json rpc requests to get txs
- Encode public key / key hash to bech32 zil address

### CODE REVIEW

File: cmd/config.go [hunk 1]
```diff
@@ -68,4 +68,6 @@ func loadDefaults() {
 	viper.SetDefault("iotex.api", "https://pharos.iotex.io/v1")
 	viper.SetDefault("waves.api", "https://nodes.wavesnodes.com")
 	viper.SetDefault("aeternity.api", "https://mdw.aepps.com/")
+	viper.SetDefault("zilliqa.api", "https://api.viewblock.io/v1/zilliqa")
```
Comment (vikmeup) [inline cmd/config.go]: + `zilliqa.api_key` ?

File: go.mod [hunk 2]
```diff
@@ -3,8 +3,14 @@ module github.com/trustwallet/blockatlas
 go 1.12.6
 
 require (
+	github.com/btcsuite/btcd v0.0.0-20190629003639-c26ffa870fd8 // indirect
+	github.com/btcsuite/btcutil v0.0.0-20190425235716-9e5f4b9a998d
```
Comment (vikmeup) [inline go.mod]: I think you only need to include `github.com/btcsuite/btcutil` here ?

File: platform/zilliqa/model.go [hunk 3]
```diff
@@ -12,3 +21,70 @@ type Tx struct {
...
+
+type TxRPC struct {
+	ID       string `json:"ID"`
+	Amount   string `json:"amount"`
+	GasLimit string `json:"gasLimit"`
+	GasPrice string `json:"gasPrice"`
+	Nonce    string `json:"nonce"`
+	Receipt  struct {
+		CumulativeGas string `json:"cumulative_gas"`
+		EpochNum      string `json:"epoch_num"`
```
Comment (vikmeup) [inline platform/zilliqa/model.go]: can you create a separate struct for `Receipt` ?

File: platform/zilliqa/model.go [hunk 4]
```diff
@@ -12,3 +21,70 @@ type Tx struct {
...
+		Success       bool   `json:"success"`
+	} `json:"receipt"`
+	SenderPubKey string `json:"senderPubKey"`
+	Signature    string `json:"signature"`
+	ToAddr       string `json:"toAddr"`
+	Version      string `json:"version"`
+}
+
+func (t *TxRPC)toTx() Tx {
```
Comment (vikmeup) [inline platform/zilliqa/model.go]: can you implement this in the api.go as `NormalizeTx` ?

File: platform/zilliqa/model.go [hunk 5]
```diff
@@ -12,3 +21,70 @@ type Tx struct {
...
+		Fee: fee.String(),
+		Signature: t.Signature,
+		Nonce: nonce,
+		ReceiptSuccess: t.Receipt.Success,
+	}
+	return tx
+}
+
+func encodePublicKeyToAddress(hexString string) string {
+	if strings.HasPrefix(hexString,"0x") {
```
Comment (vikmeup) [inline platform/zilliqa/model.go]: can you add tests for this?

File: platform/zilliqa/model.go [hunk 6]
```diff
@@ -12,3 +21,70 @@ type Tx struct {
...
+	}
+	bytes, err := hex.DecodeString(hexString)
+	if err != nil {
+		return ""
+	}
+	keyHash := sha256.Sum256(bytes)
+	return encodeKeyHashToAddress(keyHash[12:])
+}
+
+func encodeKeyHashToAddress(keyHash []byte) string {
```
Comment (vikmeup) [inline platform/zilliqa/model.go]: I would suggest having a crypto package to have this method included there?

File: platform/zilliqa/model.go [hunk 7]
```diff
@@ -12,3 +21,70 @@ type Tx struct {
...
+	conv, err := bech32.ConvertBits(keyHash, 8, 5, true)
+	if err != nil {
+		return ""
+	}
+	encoded, err := bech32.Encode("zil", conv)
```
Comment (vikmeup) [inline platform/zilliqa/model.go]: can you add `zil` as a const?

File: platform/zilliqa/model.go (hunk 4 above)
Comment (vikmeup) [reply platform/zilliqa/model.go]: + add tests for normalizing func

**Comment by vikmeup** [comment]
@hewigovens congrats with a first PR in go lang!

File: cmd/config.go (hunk 1 above)
Comment (hewigovens) [reply cmd/config.go]: api key is set in env variable

File: platform/zilliqa/model.go (hunk 3 above)
Comment (hewigovens) [reply platform/zilliqa/model.go]: This `Receipt` is used only here, I would rather keep it nested

File: platform/zilliqa/model.go (hunk 4 above)
Comment (hewigovens) [reply platform/zilliqa/model.go]: We have 3 models, `blockatlas.Tx`, `zilliqa.Tx` and `zilliqa.TxRPC`, I want to separate the concern, in `api.go` we only deal with first two.

I can add a test for converting `TxRPC` to `Tx`

**Pull request merged**